src/coreclr/vm/method.hpp

```diff
@@ -60,69 +60,83 @@ EXTERN_C VOID STDCALL PInvokeImportThunk();
-    //   - it is reversible. In particular nonconflicting signatures will map to nonconflicting ones.
-    //
-    // Async methods are called with CORINFO_CALLCONV_ASYNCCALL call convention.
+    //   - it is reversible. Thus nonconflicting signatures will map to nonconflicting ones.
     //
     // It is possible to get from one variant to another via GetAsyncOtherVariant.
     //
-    // NOTE: not all Async methods are "variants" from a pair, see AsyncExplicitImpl below.
-    //=============================================================
+    // NOTE: Not all AsyncCall methods are "variants" from a pair.
+    //       Methods that are explicitly declared as MethodImpl.Async in metadata while
+    //       not Promise-returning is a special case used in a few methods like `Await` or
+    //       other infrastructure methods used in implementation of Runtime Async itself.
+    //       Such methods do not get Task-returning facades. (We would not even know if the
+    //       facade needs to return Task or ValueTask)
+    //       Such methods can only be called from other AsyncCall methods.
+    //       Like all other cases of AsyncCall, calls to these methods have semantic of "await"
+    //       and can participate in suspension/resume.
+};
 
-    // The following methods use special calling convention (CORINFO_CALLCONV_ASYNCCALL)
-    // These methods are emitted by the JIT as resumable state machines and also take an extra
-    // parameter and extra return - the continuation object.
+inline AsyncMethodFlags operator|(AsyncMethodFlags lhs, AsyncMethodFlags rhs)
+{
+    return (AsyncMethodFlags)((int)lhs | (int)rhs);
+}
 
-    // Async methods with actual IL implementation of a MethodImpl::Async method.
-    AsyncVariantImpl,
+inline AsyncMethodFlags operator&(AsyncMethodFlags lhs, AsyncMethodFlags rhs)
+{
+    return (AsyncMethodFlags)((int)lhs & (int)rhs);
+}
 
-    // Async methods with synthetic bodies that forward to a TaskReturning method.
-    AsyncVariantThunk,
+inline AsyncMethodFlags& operator|=(AsyncMethodFlags& lhs, AsyncMethodFlags rhs)
+{
+    lhs = lhs | rhs;
+    return lhs;
+}
 
-    // Methods that are explicitly declared as Async in metadata while not Task returning.
-    // This is a special case used in a few infrastructure methods like `Await`.
-    // Such methods do not get non-Async variants/thunks and can only be called from another Async method.
-    // NOTE: These methods have the original signature and it is not possible to tell if the method is Async
-    //       from the signature alone, thus all these methods are also JIT intrinsics.
-    AsyncExplicitImpl,
-};
+inline AsyncMethodFlags& operator&=(AsyncMethodFlags& lhs, AsyncMethodFlags rhs)
+{
+    lhs = lhs & rhs;
+    return lhs;
+}
+
+inline bool hasAsyncKindFlags(AsyncMethodFlags value, AsyncMethodFlags flags)
+{
+    return (value & flags) == flags;
+}
 
 struct AsyncMethodData
 {
-    AsyncMethodKind kind;
+    AsyncMethodFlags kind;
```

Comment:
```suggestion
    AsyncMethodFlags flags;
```
Not commenting on all instances.